README.md
# Static 404

## Introduction

This extension caches your 404 page in a static HTML file, in order to save some server load.

Typo3 generates the 404 page with a loopback call. In FastCGI/php-fpm configurations where processes are limited we could face to an easy DDOS attack. Ex :
- php-fpm with pm.max_children = 50
- send 50 http concurrent request on a 404
- all processes will be locked for 30 sec. cause they loopback on the server which will wait for an available process

## What does it do?

The static 404 file is automatically generated when the page cache is cleared from the BE, then cached in typo3temp/.
When a 404 occurs, the generated file is directly send to the client without loopback call.

The plugin handles multidomain installation, and multilingual websites (if they are on separate domains).

## Installation

1.  Download, install and enable the extension in Extension manager.

2.  In the extension manager, in the configuration tab of the extension, specify the page ID of your 404 page. If you have more than one 404 page (on a multi-domain installation), you can put several IDs separated by a comma. The first ID will be used as default 404 (see below)

3.  Clear TYPO3 page cache. This will create the static 404 page in typo3temp/.

4.  In your _.htaccess file or in your virtualhost (preferable for a multi-domain installation),
    configure the ErrorDocument directive to point to the new 404 file.
    Below an excerpt of the _.htaccess file with the needed RewriteRule
    
    ```
    ErrorDocument 404 /typo3temp/tx_static404/[YOU FULL DOMAIN NAME HERE].html
    ```
    Replace `[YOU FULL DOMAIN NAME HERE]` by the domain of your installation, ie. :
    
    ```
    ErrorDocument 404 /typo3temp/tx_static404/fr.example.com.html
    ```
    
    You can also use the default 404, which corresponds to the first page ID defined :
    
    ```
    ErrorDocument 404 /typo3temp/tx_static404/default.html

## Additionnal configuration

1.  Disable pageNotFound handling : if you have an other plugin which needs to handle the pageNotfound hook, you can enable this option to delegate it. In your plugin, you will have to manage the 404 by adding this lines :

    ```
    if (t3lib_extMgm::isLoaded('static_404')) {
        $static404 = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('tx_static404');
        $static404->render404AndExit();
    }
    ```

2. Do not generate static 404 pages for domain records with a redirect

In a website with multiple sys_domain records with redirects, you can generate 404 pages for all domains by disabling this option

## Troubleshootings

If the clearCache command takes too long or never ends, check the TYPO3's log and/or the webserver's error log.
Maybe the extension is misconfigured or maybe your server can't reach itself (check your /etc/hosts file).

## Credits
&copy; 2016 GAYA La Nouvelle Agence [http://www.gaya.fr/]

ChangeLog
2016-11-10 Benoit Chenu  <contact@gaya.fr>
	* Release 1.4.1
	* Update documentation

2016-11-10 Benoit Chenu  <contact@gaya.fr>
	* Release 1.4.0
    * Unable to create file in typo3temp with long domain names
    * Add possibility to generate static pages for domains with redirection

2016-05-20 Benoit Chenu  <contact@gaya.fr>
	* Release 1.3.0
	* Change 404 file path (add tx_static404 directory)

2015-12-23 Benoit Chenu  <contact@gaya.fr>
	* Release 1.2.0
	* compatibility for TYPO3 7.6

2014-10-19 Benoit Chenu  <contact@gaya.fr>
	* Release 1.1.2
	* Remove .sxw documentation

2014-05-13 Benoit CHENU  <contact@gaya.fr>
	* release of version 1.1.1
	* small bug fixes
	* check compatibility for TYPO3 6.2

2013-09-13 Benoit CHENU  <contact@gaya.fr>
	* release of version 1.1.0
	* add disablePageNotFound_handling feature

2013-05-31 Benoit CHENU  <contact@gaya.fr>
	* release of version 1.0.0
	* initial release for TER

2013-05-10 Benoit CHENU  <contact@gaya.fr>
	* release of version 0.3.0
	* generate a default 404, useful for apache ErrorDocument directive
	* remove static declarations on methods
	* allowing class to be extended (XCLASS)

2013-02-22 Rémy DANIEL  <contact@gaya.fr>
	* release of version 0.2.0
	* better error handling
	* documentation

2013-02-21 Rémy DANIEL  <contact@gaya.fr>
	* release of version 0.1.0


ext_emconf.php
<?php

/***************************************************************
 * Extension Manager/Repository config file for ext "static_404".
 *
 * Auto generated 20-06-2013 12:30
 *
 * Manual updates:
 * Only the data in the array - everything else is removed by next
 * writing. "version" and "dependencies" must not be touched!
 ***************************************************************/

$EM_CONF[$_EXTKEY] = array (
	'title' => 'Static 404',
	'description' => 'Cache the 404 page as a static html file. See: https://github.com/agence-gaya/static_404',
	'category' => 'misc',
	'shy' => 0,
	'version' => '1.4.1',
	'dependencies' => '',
	'conflicts' => '',
	'priority' => '',
	'loadOrder' => '',
	'module' => '',
	'state' => 'stable',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearcacheonload' => 0,
	'lockType' => '',
	'author' => 'GAYA La Nouvelle Agence',
	'author_email' => 'contact@gaya.fr',
	'author_company' => 'GAYA',
	'CGLcompliance' => NULL,
	'CGLcompliance_note' => NULL,
	'constraints' => 
	array (
		'depends' => 
		array (
			'typo3' => '7.6.0-7.6.99',
		),
		'conflicts' => '',
		'suggests' => 
		array (
		),
	),
);

?>


ext_localconf.php
<?php
if (!defined ('TYPO3_MODE')) 	die ('Access denied.');

$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf'][$_EXTKEY]);

// register the clearCachePostProc hook
require_once(\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::extPath($_EXTKEY).'class.tx_static404.php');
$TYPO3_CONF_VARS['SC_OPTIONS']['t3lib/class.t3lib_tcemain.php']['clearCachePostProc'][]='tx_static404->clearCachePostProc';

if (empty($extConf['disablePageNotFound_handling'])) {
	// tell TYPO3 to serve our cached 404 when a page is not found
	$TYPO3_CONF_VARS["FE"]["pageNotFound_handling"] = 'READFILE:typo3temp/'.tx_static404::$TEMP_FILENAME_PREFIX.'/'.\TYPO3\CMS\Core\Utility\GeneralUtility::getIndpEnv('TYPO3_HOST_ONLY').'.html';
	$TYPO3_CONF_VARS["FE"]["pageNotFound_handling_statheader"] = 'HTTP/1.0 404 Not Found';
}

class.tx_static404.php
<?php
use \TYPO3\CMS\Core\Utility\GeneralUtility;
/***************************************************************
*  Copyright notice
*
*  (c) 2015 GAYA La Nouvelle Agence <contact@gaya.fr>
*  All rights reserved
*
*  This script is part of the TYPO3 project. The TYPO3 project is
*  free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 2 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * Class used for clearCachePostProc hook
 *
 * @author	Rémy Daniel <contact@gaya.fr>
 * @package	TYPO3
 * @subpackage	tx_static404
 */
class tx_static404 {

	/**
	 * Name of the temporary file stored in typo3temp
	 * If you change this name, be sure to update the rewrite rule in htaccess
	 * @var string
	 */
	static public $TEMP_FILENAME_PREFIX = 'tx_static404';

	/**
	 * This function will be called by the clearCachePostProc hook
	 *
	 * @param array $params
	 * @param object $pObj
	 * @return  void
	 */
	public function clearCachePostProc($params, $pObj) {

		/** @var \TYPO3\CMS\Core\Authentication\BackendUserAuthentication $beUser */
		$beUser =  $GLOBALS['BE_USER'];

		if ($params['cacheCmd'] === 'all' || $params['cacheCmd'] === 'pages') {
			try {
				$this->start();
			} catch (Exception $e) {
				GeneralUtility::sysLog($e->getMessage(), 'static_404', 3);
				$beUser->simplelog($e->getMessage(), 'static_404', 2);
			}
		}
	}

	/**
	 * This function will be called by the clearCachePostProc hook
	 *
	 * @throws \TYPO3\CMS\Core\Error\Exception
	 * @return  void
	 */
	private function start() {

		/** @var \TYPO3\CMS\Core\Authentication\BackendUserAuthentication $beUser */
		$beUser =  $GLOBALS['BE_USER'];

		// load the extension configuration
		$extConf = unserialize($GLOBALS['TYPO3_CONF_VARS']['EXT']['extConf']['static_404']);

		if (!is_array($extConf) || empty($extConf['all404Pids'])) {
			throw new \TYPO3\CMS\Core\Error\Exception('The extension is not properly configured. Please visit the configuration tab in the extension manager');
		}

		// extract each page id
		$pageUids = GeneralUtility::intExplode(',', $extConf['all404Pids'], true);
		if (empty($pageUids) || !$pageUids[0]) {
			throw new \TYPO3\CMS\Core\Error\Exception('Unable to find a valid 404 pid in the configuration. Please visit the configuration tab in the extension manager');
		}

		$defaultAlreadyGenerated = false;
		foreach ($pageUids as $pageUid) {
			// get all registered sys_domain for this page
			foreach ($this->allDomainRecords($pageUid) as $domain) {

				// build the url from which we will fetch the content
				$fetchUrl = $domain.'/index.php?id='.$pageUid;

				// fetch the page content
				$getUrlReport = array(
					'error' => 0,
					'message' => ''
				);
				$content = GeneralUtility::getUrl($fetchUrl, 0, false, $getUrlReport);

				if ($content !== false) {

					// write to cache file
					$urlParts = parse_url($domain);
					$host = $this->tokenizeHost($urlParts['host']);
					$tempFilename = $host.'.html';

					$error = GeneralUtility::writeFileToTypo3tempDir(PATH_site.'typo3temp/'.self::$TEMP_FILENAME_PREFIX.'/'.$tempFilename, $content);
					if (!empty($error)) {
						throw new \TYPO3\CMS\Core\Error\Exception($error);
					}

					$beUser->simplelog('Update the 404 cache for "'.$fetchUrl.'" to "'.$tempFilename.'"', 'static_404', 0);

					// The first pageUid is used as pid for the default 404
					if (!$defaultAlreadyGenerated) {
						$tempFilename = 'default.html';
						$error = GeneralUtility::writeFileToTypo3tempDir(PATH_site.'typo3temp/'.self::$TEMP_FILENAME_PREFIX.'/'.$tempFilename, $content);
						if (!empty($error)) {
							throw new \TYPO3\CMS\Core\Error\Exception($error);
						}
						$defaultAlreadyGenerated = true;
						$beUser->simplelog('Update the 404 cache for "'.$fetchUrl.'" to "'.$tempFilename.'"', 'static_404', 0);
					}

				} elseif ($getUrlReport['error'] != 0) {
					$beUser->simplelog('Error while fetching 404 page ("'.$fetchUrl.'"): '.$getUrlReport['message'], 'static_404', 2);
				}
			}
		}
	}

	/**
	 * Get all domains for a given page id.
	 * It handles the case where a page have more than one domain,
	 * which is the case on some multilingual websites.
	 * HTTPS configuration is detected both at page level or site-wide.
	 *
	 * @param  integer $pageId
	 * @return array array of domains, with protocol
	 */
	private function allDomainRecords($pageId) {

		$domains = array();

		$rootLine = \TYPO3\CMS\Backend\Utility\BackendUtility::BEgetRootLine($pageId);

		// Checks alternate domains
		if (count($rootLine) > 0) {
			// Define protocol used for this page

			/** @var \TYPO3\CMS\Frontend\Page\PageRepository $sysPage */
			$sysPage = GeneralUtility::makeInstance('TYPO3\CMS\Frontend\Page\PageRepository');
			$page = (array)$sysPage->getPage($pageId);
			$protocol = 'http';
			if ($page['url_scheme'] == 2 || ($page['url_scheme'] == 0 && GeneralUtility::getIndpEnv('TYPO3_SSL'))) {
				$protocol = 'https';
			}

			// Find every sys_domain records configured in page rootline
			foreach ($rootLine as $row) {
				$dRec = \TYPO3\CMS\Backend\Utility\BackendUtility::getRecordsByField('sys_domain', 'pid', $row['uid'], ' AND redirectTo=\'\' AND hidden=0', '', 'sorting');

				if (is_array($dRec)) {
					foreach ($dRec as $dRecord) {
						$domain = rtrim($dRecord['domainName'], '/');
						if (!$domain) {
							continue;
						}
						$domains[] = $protocol . '://' . $domain;
					}
				}
			}
		}

		// If no sys_domain found, we use TYPO3_SITE_URL
		if (count($domains) === 0) {
			$domains[] = rtrim(GeneralUtility::getIndpEnv('TYPO3_SITE_URL'), '/');
		}

		return $domains;
	}

	/**
	 * Return a tokenized version of the host.
	 * Useful to shorten the hostname.
	 * @param $host
	 * @return string
	 */
	private static function tokenizeHost($host) {
		if (strlen($host) < 50) {
			return $host;
		}

		return substr($host, 0, 40) . GeneralUtility::shortMD5($host, 10);
	}

	/**
	 * Render the 404 page
	 *
	 * Can be used by another plugin to render the 404 page
	 *
	 * @throws \RuntimeException
	 * @see disablePageNotFound_handling
	 */
	public function render404AndExit() {
		$host = self::tokenizeHost(GeneralUtility::getIndpEnv('TYPO3_HOST_ONLY'));
		$readFile = GeneralUtility::getFileAbsFileName('typo3temp/'.self::$TEMP_FILENAME_PREFIX.'/'.$host.'.html');
		if (!@is_file($readFile)) {
			throw new \TYPO3\CMS\Core\Error\Exception('Enable to find static 404 file. Try to clear Typo3 cache.');
		}
		header('HTTP/1.0 404 Not Found', null, 404);
		$fileContent = GeneralUtility::getUrl($readFile);
		$fileContent = str_replace('###CURRENT_URL###', GeneralUtility::getIndpEnv('REQUEST_URI'), $fileContent);
		$fileContent = str_replace('###REASON###', '', $fileContent);
		echo $fileContent;
		exit;
	}

}

if (defined('TYPO3_MODE') && $GLOBALS['TYPO3_CONF_VARS'][TYPO3_MODE]['XCLASS']['ext/static_404/class.tx_static404.php'])	{
	include_once($GLOBALS['TYPO3_CONF_VARS'][TYPO3_MODE]['XCLASS']['ext/static_404/class.tx_static404.php']);
}
